### PR TITLE
ORC: Change union read schema from hive to trino

### PR DESCRIPTION
**Pull request by rzhang10**
This rb changes the OC  spark reader  and vectorized reader code path to read a union data type of `[int, string]` to a spark schema/data of `[tag, field0, field1]` instead of the previous `[tag_0, tag_1]`.

### CODE REVIEW

File: orc/src/main/java/org/apache/iceberg/orc/OrcSchemaWithTypeVisitor.java [hunk 1]
```diff
@@ -71,12 +71,12 @@ protected T visitRecord(
     return visitor.record(struct, record, names, results);
   }
 
-  private static <T> T visitUnion(Type type, TypeDescription union, OrcSchemaWithTypeVisitor<T> visitor) {
+  protected T visitUnion(Type type, TypeDescription union, OrcSchemaWithTypeVisitor<T> visitor) {
```
Comment (autumnust) [inline orc/src/main/java/org/apache/iceberg/orc/OrcSchemaWithTypeVisitor.java:74]: why is this change needed?

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java [hunk 2]
```diff
@@ -311,9 +311,11 @@ private static TypeDescription buildOrcProjectForStructType(Integer fieldId, Typ
       Map<Integer, OrcField> mapping) {
     TypeDescription orcType;
     OrcField orcField = mapping.getOrDefault(fieldId, null);
+    // this branch means the iceberg struct schema actually correspond to an underlying union
     if (orcField != null && orcField.type.getCategory().equals(TypeDescription.Category.UNION)) {
       orcType = TypeDescription.createUnion();
-      for (Types.NestedField nestedField : type.asStructType().fields()) {
+      List<Types.NestedField> nestedFields = type.asStructType().fields();
```
Comment (autumnust) [inline orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java:317]: maybe precondition checking for the list size before accessing the second element and forwards?

File: spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java [hunk 3]
```diff
@@ -443,11 +445,14 @@ public ColumnVector convert(org.apache.orc.storage.ql.exec.vector.ColumnVector v
         long batchOffsetInFile) {
       UnionColumnVector unionColumnVector = (UnionColumnVector) vector;
       List<Types.NestedField> fields = structType.fields();
-      assert fields.size() == unionColumnVector.fields.length;
```
Comment (autumnust) [inline spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java:446]: Shall we preserve this kind of assertion in terms of field size?

**Review by autumnust — CHANGES_REQUESTED** [review]
Overall looks good but I still think we should gate empty union cases.

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java [hunk 4]
```diff
@@ -311,9 +311,11 @@ private static TypeDescription buildOrcProjectForStructType(Integer fieldId, Typ
       Map<Integer, OrcField> mapping) {
     TypeDescription orcType;
     OrcField orcField = mapping.getOrDefault(fieldId, null);
+    // this branch means the iceberg struct schema actually correspond to an underlying union
```
Comment (funcheetah) [inline orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java:314]: nit: `corresponds`

File: spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java [hunk 5]
```diff
@@ -115,106 +118,28 @@ public void testComplexUnion() throws IOException {
...
+      final Iterator<InternalRow> actualRowsIt = batchesToRows(reader.iterator());
 
-      Assert.assertEquals(columnarBatches.get(0).numRows(), NUM_OF_ROWS);
-      assertEquals(expectedSchema, expectedFirstRow, rowIterator.next());
-      assertEquals(expectedSchema, expectedSecondRow, rowIterator.next());
-    }
-  }
-
-  @Test
-  public void testSingleComponentUnion() throws IOException {
```
Comment (funcheetah) [inline spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java:147]: Why test for single component union is removed? How is `[ "string" ]` represented in the new schema?

File: orc/src/main/java/org/apache/iceberg/orc/OrcSchemaWithTypeVisitor.java (hunk 1 above)
Comment (funcheetah) [reply orc/src/main/java/org/apache/iceberg/orc/OrcSchemaWithTypeVisitor.java:74]: +1 on why this change

File: spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java [hunk 6]
```diff
@@ -225,23 +150,27 @@ public void testDeeplyNestedUnion() throws IOException {
 
     Schema expectedSchema = new Schema(
         Types.NestedField.optional(0, "c1", Types.StructType.of(
-            Types.NestedField.optional(1, "tag_0", Types.IntegerType.get()),
-            Types.NestedField.optional(2, "tag_1",
+            Types.NestedField.optional(100, "tag", Types.IntegerType.get()),
```
Comment (funcheetah) [inline spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java:153]: Could you explain how the value `100` is determined for the id field here?

File: orc/src/main/java/org/apache/iceberg/orc/OrcSchemaWithTypeVisitor.java (hunk 1 above)
Comment (rzhang10) [reply orc/src/main/java/org/apache/iceberg/orc/OrcSchemaWithTypeVisitor.java:74]: This is just to align with the same code pattern of other visitor's method, see `visitRecord`, also, this change enables subclass overriding, which might be necessary in the future, since this `OrcSchemaWithTypeVisitor` is a very generic visitor.

File: spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java (hunk 3 above)
Comment (rzhang10) [reply spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java:446]: I don't think it's necessary since I already know for sure the length would match, this is internal implementation detail.

File: spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java (hunk 5 above)
Comment (rzhang10) [reply spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java:147]: We already discussed before we won't support this schema and user should never use single type union cuz it doesn't make sense.

File: spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java (hunk 6 above)
Comment (rzhang10) [reply spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java:153]: it's just a random id I picked.

File: spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java (hunk 5 above)
Comment (funcheetah) [reply spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java:147]: Given it is a valid Avro schema, do we have a way to prevent user from creating this kinda of schema in the first place? What is the behavior if user do create it for current implementation?

Comment (rzhang10) [reply spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java:147]: The way to prevent it will be upon the due diligence of the user to not create such schema. We can just tell the user this case is not supported and behavior is undefined.

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java (hunk 2 above)
Comment (rzhang10) [reply orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java:317]: Same as the comment below, I don't think it's necessary here since It's already known for sure the length would be at least 2, this is an internal implementation detail.

File: spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java (hunk 5 above)
Comment (rzhang10) [reply spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java:147]: Actually, I have a second thought on this, @autumnust when gobblin writes/converts an ORC dataset and encounters an Avor `[null, string]` type, does it write just `string` type to ORC file, or does it write a union? My intuition is that you guys write a single `string` as ORC schema is nullable by itself.

Comment (autumnust) [reply spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java:147]: So Gobblin does the transformation: https://github.com/apache/gobblin/blob/5bc22a5502ddeb810a35b0fb996dd9dfc8c81121/gobblin-utility/src/main/java/org/apache/gobblin/util/orc/AvroOrcSchemaConverter.java#L62 

So a default-value representation in Avro shall not leave in a state as single-element union.

**Review by funcheetah — APPROVED** [review]
LGTM. Regarding the single item union, I think we need to verify the behavior of it in current implementation. If we do not want to support this, we need a way to document/communicate that it is not supported.
